subsys/bluetooth/mesh/key_importer.c

```diff
@@ -0,0 +1,420 @@
+/*
+ * Copyright (c) 2025 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stdlib.h>
+#include <zephyr/kernel.h>
+#include <zephyr/settings/settings.h>
+#include <zephyr/bluetooth/mesh.h>
+#include <mesh/crypto.h>
+#define LOG_LEVEL CONFIG_BT_MESH_KEYS_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(key_importer);
+
+struct net_key_legacy {
+	uint8_t unused:1,
+		kr_phase:7;
+	uint8_t val[2][16];
+} __packed;
+
+struct net_key_val {
+	uint8_t unused:1,
+		kr_phase:7;
+	struct bt_mesh_key val[2];
+} __packed;
+
+struct app_key_legacy {
+	uint16_t net_idx;
+	bool updated;
+	uint8_t val[2][16];
+} __packed;
+
+struct app_key_val {
+	uint16_t net_idx;
+	bool updated;
+	struct bt_mesh_key val[2];
+} __packed;
+
+struct net_legacy {
+	uint16_t primary_addr;
+	uint8_t dev_key[16];
+} __packed;
+
+struct net_val {
+	uint16_t primary_addr;
+	struct bt_mesh_key dev_key;
+} __packed;
+
+struct node_val_legacy {
+	uint16_t net_idx;
+	uint8_t num_elem;
+	uint8_t flags;
+	uint8_t uuid[16];
+	uint8_t dev_key[16];
+} __packed;
+
+struct node_val {
+	uint16_t net_idx;
+	uint8_t num_elem;
+	uint8_t flags;
+	uint8_t uuid[16];
+	struct bt_mesh_key dev_key;
+} __packed;
+
+static int import_net_keys(const char *path, size_t len, settings_read_cb read_cb, void *cb_arg)
+{
+	struct net_key_legacy legacy_net_key;
+	struct net_key_val net_key;
+	struct bt_mesh_key key_id;
+	uint8_t kr_phase;
+	int err;
+
+	if (len != sizeof(struct net_key_legacy)) {
+		LOG_DBG("Network key %s is not legacy", path);
+		return 0;
+	}
+
+	err = read_cb(cb_arg, &legacy_net_key, len);
+	if (err <= 0) {
+		LOG_ERR("Failed to load legacy network key %s (err %d)", path, err);
+		return err == 0 ? -EIO : err;
+	}
+
+	kr_phase = legacy_net_key.kr_phase;
+
+	LOG_DBG("kr phase %d", kr_phase);
+	LOG_HEXDUMP_DBG(&legacy_net_key.val[0][0], 16, "key one:");
+
+	err = bt_mesh_key_import(BT_MESH_KEY_TYPE_NET, &legacy_net_key.val[0][0], &key_id);
+	if (err) {
+		LOG_ERR("Failed to import first network key %s (err %d)", path, err);
+		return err;
+	}
+
+	net_key.val[0].key = key_id.key;
+
+	net_key.kr_phase = kr_phase;
+	net_key.unused = 0U;
+	if (kr_phase == BT_MESH_KR_NORMAL) {
+		net_key.val[1].key = PSA_KEY_ID_NULL;
+		goto store;
+	}
+
+	LOG_HEXDUMP_DBG(&legacy_net_key.val[1][0], 16, "key two:");
+
+	err = bt_mesh_key_import(BT_MESH_KEY_TYPE_NET, &legacy_net_key.val[1][0], &key_id);
+	if (err) {
+		LOG_ERR("Failed to import second network key %s (err %d)", path, err);
+		return err;
+	}
+
+	net_key.val[1].key = key_id.key;
+
+store:
+
+	err = settings_save_one(path, &net_key, sizeof(net_key));
+	if (err) {
+		LOG_ERR("Failed to store network key: %s (err %d)", path, err);
+		return err;
+	}
+
+	return 0;
```

Comment:
Can we add some `LOG_INF` log messages that `key number X was recovered`? Something like this?